src/lib.rs

```diff
@@ -0,0 +1,559 @@
+extern crate rand;
+
+/// A node inside a `SearchTree`.
+#[derive(Debug, Clone, PartialEq)]
+pub struct Node<T: PartialEq> {
+    pub value: T,
+    parent: Option<usize>,
+    left_child: Option<usize>,
+    right_child: Option<usize>,
+    height: usize,
+}
+
+/// A simple AVL tree.
+#[derive(Debug, PartialEq)]
+pub struct SearchTree <T: PartialEq + PartialOrd> {
+    nodes: Vec<Node<T>>,
+    root: Option<usize>,
+}
+
+impl<'a, T> SearchTree<T>
+    where T: PartialEq + PartialOrd + Clone {
+    /// Creates an empty search tree.
+    ///
+    /// # Example
+    /// ```
+    /// use avl_tree::SearchTree;
+    /// let tree: SearchTree<String> = SearchTree::new();
+    /// assert!(tree.is_empty());
+    /// ```
+    pub fn new() -> SearchTree<T> {
+        SearchTree {
+            nodes: Vec::new(),
+            root: None
+        }
+    }
+
+    /// Test if this tree is empty.
+    ///
+    /// # Example
+    /// ```
+    /// use avl_tree::SearchTree;
+    /// let mut tree: SearchTree<String> = SearchTree::new();
+    /// assert!(tree.is_empty());
+    /// tree.insert(String::from("foo"));
+    /// assert!(!tree.is_empty());
+    /// ```
+    pub fn is_empty(&self) -> bool {
+        self.root.is_none()
+    }
+
+    /// Test if this contains a given value.
+    ///
+    /// # Example
+    /// ```
+    /// use avl_tree::SearchTree;
+    /// let mut tree: SearchTree<String> = SearchTree::new();
+    /// tree.insert(String::from("hello"));
+    /// tree.insert(String::from("world!"));
+    /// assert!(tree.contains(&String::from("hello")));
+    /// assert!(!tree.contains(&String::from("cruel")));
+    /// ```
+    pub fn contains(&self, value: &T) -> bool {
+        self.get_node(value).is_ok()
+    }
+
+    /// Add the specified value to this tree.
+    ///
+    /// # Example
+    /// ```
+    /// use avl_tree::SearchTree;
+    /// let mut tree: SearchTree<String> = SearchTree::new();
+    /// tree.insert(String::from("hello"));
+    /// tree.insert(String::from("world!"));
+    /// assert!(tree.contains(&String::from("hello")));
+    /// assert!(tree.contains(&String::from("world!")));
+    /// ```
+    pub fn insert(&mut self, value: T) {
+        if let Ok(node) = self.create_node(value) {
```

Comment:
If `create_node` returns an `Err`, is the insert swallowed or ...?

Reply:
`create_node` returns an `Err` if `value` can already be found in the tree. So if `Err` is returned, insert won't do anything.

Reply:
Ah, that's a weird return type to denote a non-bad thing :) Why not have `create_node` just return a `bool` (e.g. `true` means "this node already existed")?

Reply:
Sorry, I missed this comment of yours! Currently, `insert` doesn't return anything. If the value you are trying to insert already exists in the tree, it silently fails to insert, which I think is reasonable.